contrib/packs/tests/test_action_download.py

```diff
@@ -505,6 +505,9 @@ def side_effect(ref):
         ]
 
         for default_branch, ref in edge_cases:
+            if not ref:
+                ref = PACK_INDEX['test']['version']
```

Comment:
Any insight around this test modification?

Why it changes the initial `edge_cases` scenarios list on fly that expected to be immutable?

Reply:
From my understanding, there are two StackStorm exchanges packs are used for `packs` test.
https://github.com/StackStorm-Exchange/stackstorm-test
https://github.com/StackStorm-Exchange/stackstorm-test2
They have the same pack name as `test` and is not included in `index.json` file.

This is why we have to simulate the way to get version from index.json, here is `PACK_INDEX`

Reply:
I don't think this helps. If we even place `9.9.9` as a pack version, - it will succeed as that test block tries to verify a very different thing.

---

Alright, what happens in case of reverting the code change in st2 core, will this new test fail?
If added test always succeeds irrespective of code change we focus on, how this case is intended to help?

Dig in deeper. We need a dedicated test case that will indeed cover new pack install behavior.

Reply:
This is an existing test case that new code break.

Reply:
Understood.

With this change to tests we just removed `st2 pack install test` run with no params (because it failed) replacing it with `st2 pack install test=0.4.0` run, according to 
https://github.com/StackStorm/st2/blob/429e986ddfd4bc507eb98cf6b3660ff36fd696b1/contrib/packs/tests/test_action_download.py#L535-L538

Just adding `ref=0.4.0` is not helpful there.  It's not different to the following edge case with `ref=1.2.3`;
https://github.com/StackStorm/st2/blob/429e986ddfd4bc507eb98cf6b3660ff36fd696b1/contrib/packs/tests/test_action_download.py#L496-L497

Explore other ways to fix existing tests and please include the test case that will cover new behavior in full.

Reply:
Initially,  searched index.json file and did not find `test` pack. This is why i was thinking the root cause for failure because of that. 

After do more investigation, found index is patched at https://github.com/StackStorm/st2/blob/c0458f1a06c2a81cac81697817c422ffb7631047/contrib/packs/tests/test_action_download.py#L84

Actually code comparison failed at https://github.com/StackStorm/st2/blob/c0458f1a06c2a81cac81697817c422ffb7631047/contrib/packs/tests/test_action_download.py#L523 for test case ('master', None) and ('default-branch', None). Therefore next line `return gitref` is not called. Finally it failed at `0.4.0" is not a valid version, hash, tag or branch in https://github.com/StackStorm-Exchange/stackstorm-test.`

Before fix, both `arg_ref` and `ref` are None. With fix, `arg_ref`=`0.4.0` and still is `ref=None`

Reply:
Good stuff! :+1:

Reply:
Good catch! Thanks!